drivers/sensor/vcnl4040/vcnl4040.c

```diff
@@ -0,0 +1,358 @@
+
+	if (vcnl4040_read(dev, VCNL4040_REG_ALS_DATA, &data->light)) {
+		LOG_ERR("Could not fetch ambient light");
+		ret = -EIO;
+		goto exit;
+	}
+	/*
+	 * scale the lux depending on the value of the integration time
+	 * see page 8 of the VCNL4040 application note:
+	 * https://www.vishay.com/docs/84307/designingvcnl4040.pdf
+	 */
+	data->light = (10 * data->light) >> config->als_it;
+#endif
+exit:
+	k_sem_give(&data->sem);
+
+	return ret;
+}
+
+static int vcnl4040_channel_get(const struct device *dev,
+				enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct vcnl4040_data *data = dev->data;
+	int ret = 0;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	switch (chan) {
+	case SENSOR_CHAN_PROX:
+		val->val1 = data->proximity;
+		val->val2 = 0;
+		break;
+
+#ifdef CONFIG_VCNL4040_ENABLE_ALS
+	case SENSOR_CHAN_LIGHT:
+		val->val1 = data->light;
+		val->val2 = 0;
+		break;
+#endif
+
+	default:
+		ret = -ENOTSUP;
+	}
+
+	k_sem_give(&data->sem);
+
+	return ret;
+}
+
+static int vcnl4040_proxy_setup(const struct device *dev)
+{
+	const struct vcnl4040_config *config = dev->config;
+	uint16_t conf = 0;
+
+	if (vcnl4040_read(dev, VCNL4040_REG_PS_MS, &conf)) {
+		LOG_ERR("Could not read proximity config");
+		return -EIO;
+	}
+
+	/* Set LED current */
+	conf |= config->led_i << VCNL4040_LED_I_POS;
+
+	if (vcnl4040_write(dev, VCNL4040_REG_PS_MS, conf)) {
+		LOG_ERR("Could not write proximity config");
+		return -EIO;
+	}
+
+	if (vcnl4040_read(dev, VCNL4040_REG_PS_CONF, &conf)) {
+		LOG_ERR("Could not read proximity config");
+		return -EIO;
+	}
+
+	/* Set PS_HD */
+	conf |= VCNL4040_PS_HD_MASK;
+	/* Set duty cycle */
+	conf |= config->led_dc << VCNL4040_PS_DUTY_POS;
+	/* Set integration time */
+	conf |= config->proxy_it << VCNL4040_PS_IT_POS;
+	/* Clear proximity shutdown */
+	conf &= ~VCNL4040_PS_SD_MASK;
+
+	if (vcnl4040_write(dev, VCNL4040_REG_PS_CONF, conf)) {
+		LOG_ERR("Could not write proximity config");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_VCNL4040_ENABLE_ALS
+static int vcnl4040_ambient_setup(const struct device *dev)
+{
+	const struct vcnl4040_config *config = dev->config;
+	uint16_t conf = 0;
+
+	if (vcnl4040_read(dev, VCNL4040_REG_ALS_CONF, &conf)) {
+		LOG_ERR("Could not read proximity config");
+		return -EIO;
+	}
+
+	/* Set ALS integration time */
+	conf |= config->als_it << VCNL4040_ALS_IT_POS;
+	/* Clear ALS shutdown */
+	conf &= ~VCNL4040_ALS_SD_MASK;
+
+	if (vcnl4040_write(dev, VCNL4040_REG_ALS_CONF, conf)) {
+		LOG_ERR("Could not write proximity config");
+		return -EIO;
+	}
+
+	return 0;
+}
+#endif
+
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+static int vcnl4040_device_ctrl(const struct device *dev,
+				uint32_t ctrl_command, void *context,
+				device_pm_cb cb, void *arg)
+{
+	int ret = 0;
+
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+		uint32_t device_pm_state = *(uint32_t *)context;
+		uint16_t ps_conf;
+
+		if (vcnl4040_read(dev, VCNL4040_REG_ALS_CONF, &ps_conf)) {
+			return -EIO;
```

Comment:
I think it's better to propagate error codes, unless `-EIO` conversion is intended.

Reply:
I agree, changed